### PR TITLE
feat: support `chat_template_kwargs` in tokenizer config

### PR DESCRIPTION
**Pull request by xxman-google**
# What does this PR do ?

**Support additional arg `chat_template_kwargs` inside tokenizer config**

This is useful for models that have both `thinking` and `non-thinking` modes.

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)): N/A

# Usage
* **Run evaluation using Qwen3 under thinking mode**

```python
uv run python examples/run_eval.py --config examples/configs/evals/qwen3_thinking.yaml
```
* **Run evaluation using Qwen3 under non-thinking mode**

```python
uv run python examples/run_eval.py --config examples/configs/evals/qwen3_non_thinking.yaml
```
Also, verified that the metric is much higher using `thinking` mode. Made sure that `run_sft.py`, `run_dpo.py`, `run_grpo_math.py` are working.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [x] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

**Comment by terrykong** [comment]
converting to draft since @xxman-google is still working. feel free to convert from draft when ready

File: examples/configs/evals/aime2024_non_thinking.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  model_name: "Qwen/Qwen3-8B"
+  temperature: 0.7
+  top_p: 0.8
+  top_k: 20
+  vllm_cfg:
+    max_model_len: 8192
+
+tokenizer:
+  chat_template_kwargs:
+    enable_thinking: false
```
Comment (yuki-97) [inline examples/configs/evals/aime2024_non_thinking.yaml]: Do you mind adding this param to `examples/configs/evals/eval.yaml` and set it default to false?
And not adding these two files: `examples/configs/evals/aime2024_non_thinking.yaml` and `examples/configs/evals/aime2024_thinking.yaml`.

File: docs/guides/eval.md [hunk 2]
```diff
@@ -48,6 +48,12 @@ uv run python examples/run_eval.py
 # Run evaluation script with converted model
 uv run python examples/run_eval.py generation.model_name=$PWD/results/grpo/hf
 
+# Run evaluation script with Qwen3 model under thinking mode
```
Comment (yuki-97) [inline docs/guides/eval.md:51]: Do you mind only keeping the thinking mode example but by this way?
```bash
uv run python examples/run_eval.py \
    generation.model_name=Qwen/Qwen3-8B \
    generation.temperature=0.6 \
    generation.top_p=0.95 \
    generation.top_k=20 \
    generation.vllm_cfg.max_model_len=32768 \
    tokenizer.chat_template_kwargs.enable_thinking=true \
    data.dataset_name=aime2024 \
    data.prompt_file=examples/prompts/cot.txt
```

File: examples/run_dpo.py [hunk 3]
```diff
@@ -186,10 +187,12 @@ def setup_data(data_config: DataConfig, policy_config: PolicyConfig):
 
     dpo_task_spec = data.task_spec
 
-    tokenizer = get_tokenizer(policy_config["tokenizer"])
+    tokenizer_config = policy_config["tokenizer"]
+    tokenizer = get_tokenizer(tokenizer_config)
     train_dataset = AllTaskProcessedDataset(
         train_dataset,
         tokenizer,
+        tokenizer_config,
```
Comment (yuki-97) [inline examples/run_dpo.py]: Could you only pass `tokenizer_config["chat_template_kwargs"]` to make it clear? Since we don't need other things in `tokenizer_config`.

File: examples/run_grpo_math.py
Comment (yuki-97) [inline examples/run_grpo_math.py:1]: `examples/run_grpo_sliding_puzzle.py` also needs to be updated.

File: tests/unit/data/test_llm_message_utils.py
Comment (yuki-97) [inline tests/unit/data/test_llm_message_utils.py:1]: Could you add an unit test to guard task_spec and tokenizer could work well w/ and w/o enable_thinking?
Ideally they should have different output.

File: docs/guides/eval.md (hunk 2 above)
Comment (xxman-google) [reply docs/guides/eval.md:51]: Done

File: examples/configs/evals/aime2024_non_thinking.yaml (hunk 1 above)
Comment (xxman-google) [reply examples/configs/evals/aime2024_non_thinking.yaml]: Done

File: examples/run_dpo.py (hunk 3 above)
Comment (xxman-google) [reply examples/run_dpo.py]: Done

File: tests/unit/data/test_llm_message_utils.py
Comment (xxman-google) [reply tests/unit/data/test_llm_message_utils.py:1]: added two tests,
`test_get_formatted_message_log_add_generation_prompt_qwen3_disable_thinking`
`test_get_formatted_message_log_add_generation_prompt_qwen3_enable_thinking`.

File: examples/run_grpo_math.py
Comment (xxman-google) [reply examples/run_grpo_math.py:1]: Done

File: tests/unit/data/test_llm_message_utils.py
Comment (yuki-97) [reply tests/unit/data/test_llm_message_utils.py:1]: Thanks @xxman-google , do you mind combining them into one unit test?
Feels the two tests are almost the same except `enable_thinking` and which `chat_message_log` to use.

I think we could use `enable_thinking` as a param, and decide which `chat_message_log` to use in the unit test, then use `@pytest.mark.parametrize("enable_thinking", [True, False])` to combine them into one unit test.

File: examples/run_dpo.py [hunk 4]
```diff
@@ -186,10 +187,12 @@ def setup_data(data_config: DataConfig, policy_config: PolicyConfig):
 
     dpo_task_spec = data.task_spec
 
-    tokenizer = get_tokenizer(policy_config["tokenizer"])
+    tokenizer_config = policy_config["tokenizer"]
+    tokenizer = get_tokenizer(tokenizer_config)
     train_dataset = AllTaskProcessedDataset(
         train_dataset,
         tokenizer,
+        tokenizer_config.get("chat_template_kwargs", dict()),
```
Comment (yuki-97) [inline examples/run_dpo.py]: One more thing is that could you change all the `dict()` to `{}`?

Comment (xxman-google) [reply examples/run_dpo.py]: Done

File: tests/unit/data/test_llm_message_utils.py
Comment (xxman-google) [reply tests/unit/data/test_llm_message_utils.py:1]: Done

**Comment by xxman-google** [comment]
@yuki-666 comments addressed. PTAL.

File: tests/unit/data/test_data_processor.py [hunk 5]
```diff
@@ -51,6 +51,7 @@ def test_math_data_processor():
     dataset = AllTaskProcessedDataset(
         dataset=raw_dataset,
         tokenizer=tokenizer,
+        chat_template_kwargs=dict(),
```
Comment (yuki-97) [inline tests/unit/data/test_data_processor.py]:
```suggestion
        chat_template_kwargs={},
```
missing here.

Comment (xxman-google) [reply tests/unit/data/test_data_processor.py]: @yuki-666  Done. Also, fixed `run_rm.py` since it is a new file.

**Review by yuki-666 — APPROVED** [review]
Thanks @xxman-google , LGTM! Approved.

@terrykong Can you also take a review when you have time?

**Review by terrykong — CHANGES_REQUESTED** [review]
thanks for adding this feature

Could you add this in the typeddict for tokenizerconfig https://github.com/NVIDIA-NeMo/RL/blob/56040244e94bb59f70c4b83732175821306bb09b/nemo_rl/models/policy/__init__.py#L102

@yuki-666 what do you think about just plumbing the chat template args via partial(). it'll probably lead to a simpler API since we don't have to plumb the args everywhere

**Comment by yuki-666** [comment]
> @yuki-666 what do you think about just plumbing the chat template args via partial(). it'll probably lead to a simpler API since we don't have to plumb the args everywhere

that's a good idea! then we only need to set it once.

**Comment by terrykong** [comment]
@xxman-google what do you think about the proposal of setting these args in a functools.partial and overwriting it on the tokenizer to avoid having to plumb the kwarg everywhere?

File: examples/configs/evals/gpqa_eval.yaml [hunk 6]
```diff
@@ -12,4 +12,4 @@ data:
 
 env:
   math:
-    verifier_type: "multichoice"
+    verifier_type: "english_multichoice"
```
Comment (yuki-97) [inline examples/configs/evals/gpqa_eval.yaml:15]: for `examples/configs/evals/gpqa_eval.yaml`, `examples/configs/evals/mmlu.yaml`, `examples/configs/evals/mmlu_zh_cn.yaml`

should all these to be `multilingual_multichoice`? and we can remove setting `verifier_type` in `mmlu_zh_cn` since it's inherit from `mmlu`.

as far as I known, `english_multichoice` detects 10 options (A-J) in english (it is mainly used for MMLU-Pro), and `multilingual_multichoice` detects 4 options (A-D) in multilingual (so it's better to use it if the answer is 4 options). please correct me if my understanding is not correct.

**Comment by yuki-97** [comment]
hi @xxman-google , since the below two eval configs are broken at main for a long time, I split the fix to https://github.com/NVIDIA-NeMo/RL/pull/1123 and add you as a co-author.
```bash
uv run python examples/run_eval.py --config examples/configs/evals/gpqa_eval.yaml
uv run python examples/run_eval.py --config examples/configs/evals/mmlu.yaml
```

for the `chat_template_kwargs` support, do you mind to update them or I can help to update them, wdyt?

**Comment by euronymous-aithal** [comment]
@yuki-97  please review

**Comment by yuki-97** [comment]
> @yuki-97 please review

yea, I'll create another PR later to involve the changes in this PR and address the remaining comments.
